File: bench.cpp
#include <chrono>
#include <cstdio>
#include <memory>
#include <utility>

struct Node {
    int value;
    std::unique_ptr<Node> next;
};

class LinkedList {
public:
    void insert(int v) {
        auto n = std::make_unique<Node>();
        n->value = v;
        n->next = std::move(head_);
        head_ = std::move(n);
    }

    bool find(int v) const {
        const Node* cur = head_.get();
        while (cur) {
            if (cur->value == v) return true;
            cur = cur->next.get();
        }
        return false;
    }

    ~LinkedList() {
        auto cur = std::move(head_);
        while (cur) {
            auto next = std::move(cur->next);
            cur = std::move(next);
        }
    }

private:
    std::unique_ptr<Node> head_;
};

int main() {
    using clock = std::chrono::steady_clock;
    using ms = std::chrono::duration<double, std::milli>;

    const int N = 1'000'000;
    const int M = 50;

    std::printf("C++  N=%d full-traversal misses M=%d\n", N, M);
    for (int trial = 1; trial <= 3; ++trial) {
        auto list = std::make_unique<LinkedList>();

        auto t0 = clock::now();
        for (int i = 0; i < N; ++i) list->insert(i);
        auto t1 = clock::now();

        volatile int hits = 0;
        for (int j = 0; j < M; ++j) if (list->find(-1)) hits++;
        auto t2 = clock::now();

        list.reset();
        auto t3 = clock::now();

        std::printf("  trial %d  build=%7.1f ms  find=%7.1f ms  drop=%7.1f ms\n",
                    trial,
                    ms(t1 - t0).count(),
                    ms(t2 - t1).count(),
                    ms(t3 - t2).count());
    }
    return 0;
}

File: bench.rs
use std::hint::black_box;
use std::time::Instant;

struct Node {
    value: i32,
    next: Option<Box<Node>>,
}

pub struct LinkedList {
    head: Option<Box<Node>>,
}

impl LinkedList {
    pub fn new() -> Self {
        LinkedList { head: None }
    }

    pub fn insert(&mut self, value: i32) {
        self.head = Some(Box::new(Node {
            value,
            next: self.head.take(),
        }));
    }

    pub fn find(&self, value: i32) -> bool {
        let mut current = &self.head;
        while let Some(node) = current {
            if node.value == value {
                return true;
            }
            current = &node.next;
        }
        false
    }
}

impl Drop for LinkedList {
    fn drop(&mut self) {
        let mut current = self.head.take();
        while let Some(mut node) = current {
            current = node.next.take();
        }
    }
}

fn main() {
    const N: i32 = 1_000_000;
    const M: i32 = 50;

    println!("Rust N={N} full-traversal misses M={M}");
    for trial in 1..=3 {
        let mut list = Box::new(LinkedList::new());

        let t0 = Instant::now();
        for i in 0..N {
            list.insert(i);
        }
        let t1 = Instant::now();

        let mut hits: u32 = 0;
        for _ in 0..M {
            if list.find(-1) {
                hits += 1;
            }
        }
        black_box(hits);
        let t2 = Instant::now();

        drop(list);
        let t3 = Instant::now();

        let build = t1.duration_since(t0).as_secs_f64() * 1000.0;
        let find = t2.duration_since(t1).as_secs_f64() * 1000.0;
        let dropt = t3.duration_since(t2).as_secs_f64() * 1000.0;
        println!("  trial {trial}  build={build:7.1} ms  find={find:7.1} ms  drop={dropt:7.1} ms");
    }
}
